### PR TITLE
'palette' input type

### PR DESCRIPTION
**Pull request by mousewax**
Added a simplified 'palette' input type that takes a list of colors as options. 

```
        {
          "type": "palette",
          "label": "Color",
          "options": [
            "#897349",
            "#89dd98",
            "#fff",
            "#dbdc00"
          ],
          "variable": "playerColor",
          "value": "${playerColor}"
        }
```
image

It uses --VTTblue as the selected outline and checkmark color.

----------------
Proposed wiki update:

* `type`: 
......
    * `palette`_displays colored radio buttons_-stores the selected value. Unlike the `color` field, this offers a consistent viewing experience across browsers and forces the players to choose from a palette of options.
      * `colors`: _array_ - displays the color choices available to the player. Accepts most color formats (html-friendly, hex, rgb, rgba, hsl, hsla).

### CODE REVIEW

**Comment by ArnoldSmith86** [comment]
PR-SERVER-BOT: You can play around with it here: https://test.virtualtabletop.io/PR-1625/pr-test (or any other room on that server)

After merging, a backup will be available at `https://beta.virtualtabletop.io/editor/PR1625-pr-test`.

**Comment by 96LawDawg** [comment]
> It uses --VTTblue as the selected outline and checkmark color.

That won't work when you have a dark blue playerColor.  Use the colorContrast function.  Maybe I can figure out how to make that happen.  I'll look.

**Comment by robartsd** [comment]
> > It uses --VTTblue as the selected outline and checkmark color.
> 
> That won't work when you have a dark blue playerColor. Use the colorContrast function. Maybe I can figure out how to make that happen. I'll look.

There is still the white check mark, so even if the color is exactly --VTTblue there will still be some indication. Tricky part is that it should have adequate contrast between both the selected color and white.

**Comment by 96LawDawg** [comment]
That's the magic of the colorContrast function.

**Comment by 96LawDawg** [comment]
I can't figure it out.  When nothing worked, to dumb it down, I tried just putting a simple console.log() into domhelper.js where your code is.  Never worked.  domhelper must do something different than I'm used to.  Here's what I was trying to do:
 - Get the currently selected field.value and determine its color and put it into some variable, say "contrast."
 - Call colorContrast(contrast).
 - Instead of --VTTBlue, use the result of the colorContrast function.  As is with the default parameters you get white or black, but you could also pass a value in the function to make it some shade of a similar hue.
 - Then make the checkmark color itself the same as the original.

That was my idea any way.  Because this doesn't have quite the same zing:
Blue

**Comment by mousewax** [comment]
I don't get what you mean by `options` is not working right. It seems to be working fine to create radio inputs. Am I missing something?

**Comment by 96LawDawg** [comment]
Just the JSON Editor default stuff is not right.

When you use the JSON Editor to select the palette field and then press the "options" button you get:

```
       {
          "type": "palette",
          "options": [
            {
              "value": "value",
              "text": "text"
            }
          ]
        }
 ``

Because that is set up for the "select" type field.  Options should just be an empty array:  [    ].  I can't get that to work.

**Comment by mousewax** [comment]
Input elements and UI should generally be consistent so I added a contrasty border rather than change the color.

image

**Comment by 96LawDawg** [comment]
> Input elements and UI should generally be consistent so I added a contrasty border rather than change the color.

That works, is consistent, and doesn't require all the code shenanigans I had come up with.

One thing doesn't work:  Getting the color value out of the INPUT function.  It just returns null.

**Comment by robartsd** [comment]
> I made the JSON Editor mostly work with the new palette field. However, options doesn't work because it shows the options for the select field. I can't figure out how to get it to show something different. The easiest way would be to change "options" to something else like "colors."

I'm in favor of this change

```
        {
          "type": "palette",
          "label": "Color",
          "colors": [
            "#897349",
            "#89dd98",
            "#fff",
            "#dbdc00"
          ],
          "variable": "playerColor",
          "value": "${playerColor}"
        }
```

**Comment by mousewax** [comment]
Unfortunately I am not sure why it's passing null. The HTML looks fine and I guess I expected it to pass right through like any other form element.

**Comment by robartsd** [comment]
> Unfortunately I am not sure why it's passing null. The HTML looks fine and I guess I expected it to pass right through like any other form element.

The code for INPUT reads the input from the DOM. I don't think it knows how to read radio button inputs yet.

**Comment by mousewax** [comment]
> The code for INPUT reads the input from the DOM. I don't think it knows how to read radio button inputs yet. Widget.evaluateInputOverlay() needs to be updated to use the new input.

Thanks! I think I did it correctly.

**Comment by robartsd** [comment]
> > The code for INPUT reads the input from the DOM. I don't think it knows how to read radio button inputs yet. Widget.evaluateInputOverlay() needs to be updated to use the new input.
> 
> Thanks! I think I did it correctly.

I'd use .filter() or .reduce() rather than a for loop, but this should work.

**Review by robartsd — APPROVED** [review]

**Review by ArnoldSmith86 — APPROVED** [review]
Code looks fine. I didn't test anything.

**Comment by mousewax** [comment]
> I'd use .filter() or .reduce() rather than a for loop, but this should work.

I don't quite understand how to use those effectively without still looping over the smaller array. I'd love to see how they work in practice. I am just not that smart.

**Comment by mousewax** [comment]
In the `colors` section it will accept any html-friendly color (like "red" or "blue") but the system converts them to hex colors so it will not re-check the selected color when you re-open the input.  Not the end of the world, but it should probably be noted in the wiki.

**Comment by ArnoldSmith86** [comment]
We can compare to `toHex(color)`.

**Comment by RaphaelAlvez** [comment]
https://user-images.githubusercontent.com/73476656/219034109-a7bdcb9d-c1c2-4caf-af80-7f37d2dd8534.mp4

Minor visual bug when takes some time to load the font. Testing it on my phone on the street so maybe we should just ignore it.

**Comment by mousewax** [comment]
I noticed that on the PR test server as well. It will be the same with any other element where the font has not yet loaded.

**Comment by mousewax** [comment]
> We can compare to `toHex(color)`.

I didn't realize that was a thing. It's been added.

**Review by 96LawDawg — APPROVED** [review]
Approved through test.

However, there is one small change I request if possible.  Something about the rounding of the corners and the white border is not quite right.  You can see small bits of color peeking out at the corners.  Can you try to fix that?

Corners

**Comment by 96LawDawg** [comment]
I added a proposed wiki update at the very top.  Also tested with hsl, hsla, rgb, and rgba and they all work fine.

**Comment by 96LawDawg** [comment]
> I believe I have now fixed this.

Yes.  Should be ready to merge this in 24 hours.

**Comment by robartsd** [comment]
> I added a proposed wiki update at the very top. Also tested with hsl, hsla, rgb, and rgba and they all work fine.

Any way the browser running the code supports that results in an sRGB color should work. Not sure what happens in browsers that support colors outside of sRGB (I believe only Safari so far) if a color outside sRGB is specified (preferred behavior would be the closest sRGB color, may get black).

**Comment by 96LawDawg** [comment]
> Any way the browser running the code supports that results in an sRGB color should work. Not sure what happens in browsers that support colors outside of sRGB (I believe only Safari so far) if a color outside sRGB is specified (preferred behavior would be the closest sRGB color, may get black).

I'm pretty sure toHex returns black in that situation.  I don't think we got that far in the color function PR.

**Comment by robartsd** [comment]
> > Any way the browser running the code supports that results in an sRGB color should work. Not sure what happens in browsers that support colors outside of sRGB (I believe only Safari so far) if a color outside sRGB is specified (preferred behavior would be the closest sRGB color, may get black).
> 
> I'm pretty sure toHex returns black in that situation. I don't think we got that far in the color function PR.

Yes we use the fillstyle property of a canvas element to evaluate colors. Safari (and Chrome > 110) support the `color()` function (but it is possible that it returns colors outside sRGB), but Firefox does not. All supported browsers should support `hwb()` color function (which also always returns sRGB).
